contracts/signature/Eip712Signature.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity >=0.8.0 <0.9.0;
+
+/// @title EIP712Signature - A contract that extracts and inspects EIP-712 signatures appended to calldata.
+
+abstract contract EIP712Signature {
+    uint256 private nonce;
+
+    /**
+     * @dev Returns the current nonce value.
+     */
+    function eip712Nonce() public view returns (uint256) {
+        return nonce;
+    }
+
+    /**
+     * @dev Increments nonce.
+     */
+    function eip712BumpNonce() internal {
+        nonce = nonce + 1;
+    }
+
+    /**
+     * @dev When signature present in calldata, returns the address of the signer.
+     */
+    function eip712SignedBy() internal returns (address signer) {
+        if (msg.data.length >= 4 + 32 + 32 + 1) {
+            (
+                bytes calldata dataTrimmed,
+                uint8 v,
+                bytes32 r,
+                bytes32 s
+            ) = _splitSignature(msg.data);
+
+            bytes32 txHash = _moduleTxHash(dataTrimmed, nonce);
+
+            signer = ecrecover(txHash, v, r, s);
+        }
+    }
+
+    /**
+     * @dev Extracts signature from calldata, and divides it into `uint8 v, bytes32 r, bytes32 s`.
+     * @param data current transaction calldata.
+     */
+    function _splitSignature(
+        bytes calldata data
+    )
+        private
+        pure
+        returns (bytes calldata dataTrimmed, uint8 v, bytes32 r, bytes32 s)
+    {
+        uint256 length = data.length;
+        dataTrimmed = data[0:length - 65];
```

Comment:
Also, is there any guarantees that actual data that was supposed to be signed by the user is located at `msg.data[0:length - 65]`?
I mean, I don't see the exact strange case now, but generally this code expects "normal" case where the msg.data was just elongationed with `r,s,v`, but theoretically (for example with some non-abi-standard packing of original `to,value,data` parameters) `msg.data[length - 65:]` can intersect with the real data signed by the user.

Reply:
Right, if no signature was appended, we are going to recover to some random address that was anyway not enabled as module